### PR TITLE
Prevent Mantid crash when fitting Gaussian peaks for a CrystalFieldFunction

### LINKED ISSUE #34245: Mantid crash when fitting Gaussian peaks in CEF
**Describe the bug**
Mantid crashes when fitting Gaussian peaks for a multisite CrystalField object. With the default peak shape, Lorentzian, this problem does not occur.

**To Reproduce**
Tun the following script:

`from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import re

from CrystalField.fitting import makeWorkspace
from CrystalField import CrystalField, CrystalFieldMultiSite, CrystalFieldFit, Background, Function

c_mbsr = 79.5774715459  # Conversion from barn to mb/sr
origin = CrystalField('Ce', 'C2v', B20=0.37737, B22=3.9770, B40=-0.031787, B42=-0.11611, B44=-0.12544, Temperature=44.0)#, FWHM=1.1)
x, y = origin.getSpectrum()
ws = makeWorkspace(x, y)
               
params = {'B20': 0.37737, 'B22': 3.9770, 'B40': -0.031787, 'B42': -0.11611, 'B44': -0.12544,
                  #'Temperature': 44.0, 'FWHM': 1.1}
                  'Temperature': [44.0, 50.0], 'FWHM': [1.1, 1.6]}
cf1 = CrystalField('Ce', 'C2v', **params)
cf2 = CrystalField('Pr', 'C2v', **params)
cf = cf1 + cf2

cf.PeakShape = 'Gaussian'

fit = CrystalFieldFit(Model=cf, InputWorkspace=[ws,ws],MaxIterations=10,Minimizer='BFGS')

fit.fit()`

### PR DESCRIPTION
**Pull request by SilkeSchomann**
**Description of work.**

Added check for height ties based on sigma for Gaussian peaks within a CrystalFieldFunction to void IFunction::sortTies(). In this case the tie is not sorted but still applied as usual for its respective peak function. The sortTies function does not differentiate between ties applied to individual peaks separately and ties that can potentially influence each other. This caused Mantid crashes for multisite CrystalField objects with Gaussian peaks as Gaussian peaks automatically have a tie based on sigma for its height.

**To test:**

Run the script from the original issue [34245](https://github.com/mantidproject/mantid/issues/34245) and check it works.

Fixes #34245 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by MialLewis — DISMISSED** [review]
Script that previously crashed mantid runs with a successful fit, code looks good.

I note that several `Numeric derivatives should be used when non-constant ties defined.` warnings are output during the fit (at least on `RelWithDebug` mode), probably a non-issue.

File: Framework/API/src/IFunction.cpp [hunk 1]
```diff
@@ -1584,6 +1584,12 @@ void IFunction::sortTies() {
     if (!tie) {
       continue;
     }
+    // Ignore height ties for Gaussian peaks as component of a CrystalFieldFunction
+    if (this->name() == "CrystalFieldFunction" && tie->ownerFunction()->name() == "Gaussian" &&
```
Comment (martyngigg) [inline Framework/API/src/IFunction.cpp:1588]: Could I ask if there would be a different way to solve this that doesn't required quite detailed knowledge of the function/tie specifics in the base class?

Maybe an `bool IFunction::ignoreTie(const Tie&)` that a function could override that by default returned `false`?

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — CHANGES_REQUESTED** [review]
Thanks for these changes. It's good that functions can now decide this for themselves.

Would it be possible to add a test for `IFunction::sortTies` to check those that are expected to be ignored are ignored?

**Comment by SilkeSchomann** [comment]
> 

The result of sortTies is not accessible. I've added tests for ignoreTie instead and a Python test that would have caught the initial problem.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**